src/interface/viewport.c

```diff
@@ -1398,14 +1398,47 @@ void viewport_vehicle_paint_setup(rct_vehicle *vehicle, int imageDirection)
 }
 
 /**
- *
- *  rct2: 0x0068F0FB
+ * rct2: 0x0068F0FB
  */
-void viewport_peep_paint_setup(rct_peep *peep, int imageDirection)
+void viewport_peep_paint_setup(rct_peep * peep, int imageDirection)
 {
-	RCT2_CALLPROC_X(0x0068F0FB, peep->x, imageDirection, peep->y, peep->z, (int)peep, 0, 0);
-}
+	rct_drawpixelinfo * dpi = RCT2_GLOBAL(0x140E9A8, rct_drawpixelinfo*);
+	if (dpi->zoom_level > 2) {
+		return;
+	}
+
+	if (gCurrentViewportFlags & VIEWPORT_FLAG_INVISIBLE_PEEPS) {
+		return;
+	}
+
+	rct_sprite_entry sprite = g_sprite_entries[peep->sprite_type];
+
+	uint8 spriteType = peep->action_sprite_type;
+	uint8 imageOffset = peep->action_sprite_image_offset;
+
+	if (peep->action == PEEP_ACTION_NONE_1) {
+		spriteType = peep->next_action_sprite_type;
+		imageOffset = 0;
+	}
 
+	uint32 imageId = (imageDirection >> 3) + sprite.sprite_image[spriteType].base_image + imageOffset * 4;
```

Comment:
call this `baseImageId` and create an `imageId` for each function that way it wont be such a ridiculously long function call.